ui/job-view/context/Pushes.jsx

```diff
@@ -0,0 +1,408 @@
+  }
+
+  addPushes(data) {
+    const { $rootScope } = this.props;
+    const { pushList } = this.state;
+
+    if (data.results.length > 0) {
+      const pushIds = pushList.map(push => push.id);
+      pushList.push(...data.results.filter(push => !pushIds.includes(push.id)));
+      pushList.sort((a, b) => b.push_timestamp - a.push_timestamp);
+      const oldestPushTimestamp = pushList[pushList.length - 1].push_timestamp;
+      this.recalculateUnclassifiedCounts();
+      this.setValue(
+        { pushList: [...pushList], oldestPushTimestamp },
+        () => this.setRevisionTips(pushList),
+      );
+      $rootScope.firstPush = pushList[0];
+      $rootScope.$apply();
+    }
+  }
+
+  updateUrlFromchange() {
+    // since we fetched more pushes, we need to persist the push state in the URL.
+    const { pushList } = this.state;
+    const updatedLastRevision = pushList[pushList.length - 1].revision;
+
+    if (getUrlParam('fromchange') !== updatedLastRevision) {
+      this.skipNextPageReload = true;
+      setUrlParam('fromchange', updatedLastRevision);
+    }
+  }
+
+  /**
+   * Loops through the map of unclassified failures and checks if it is
+   * within the enabled tiers and if the job should be shown. This essentially
+   * gives us the difference in unclassified failures and, of those jobs, the
+   * ones that have been filtered out
+   */
+  recalculateUnclassifiedCounts() {
+    const { jobMap } = this.state;
+    const { filterModel, $rootScope } = this.props;
+    const tiers = filterModel.urlParams.tier || [];
+    let allUnclassifiedFailureCount = 0;
+    let filteredUnclassifiedFailureCount = 0;
+
+    Object.values(jobMap).forEach((job) => {
+      if (isUnclassifiedFailure(job)) {
+        if (tiers.includes(String(job.tier))) {
+          if (filterModel.showJob(job)) {
+            filteredUnclassifiedFailureCount++;
+          }
+          allUnclassifiedFailureCount++;
+        }
+      }
+    });
+    this.setValue({ allUnclassifiedFailureCount, filteredUnclassifiedFailureCount });
+    $rootScope.unclassifiedFailureCount = allUnclassifiedFailureCount;
+    $rootScope.$apply();
+  }
+
+  updateJobMap(jobList) {
+    const { jobMap, pushList } = this.state;
+
+    if (jobList.length) {
+      const push = pushList.find(push => push.id === jobList[0].push_id);
+      push.jobsLoaded = true;
+      // lodash ``keyBy`` is significantly faster than doing a ``reduce``
+      this.setValue({
+        jobMap: { ...jobMap, ...keyBy(jobList, 'id') },
+        jobsLoaded: pushList.every(push => push.jobsLoaded),
+        pushList: [...pushList],
+      });
+    }
+  }
+
+  render() {
+    return (
+      <PushesContext.Provider value={this.value}>
+        {this.props.children}
+      </PushesContext.Provider>
+    );
+  }
+
+}
+
+Pushes.propTypes = {
+  children: PropTypes.object.isRequired,
+  filterModel: PropTypes.object.isRequired,
+  $rootScope: PropTypes.object.isRequired,
+  thNotify: PropTypes.object.isRequired,
```

Comment:
During the tests, this warning is generated:

```
'Warning: Failed prop type: The prop `thNotify` is marked as required in `Pushes`, but its value is `undefined`.
    in Pushes'
```

Reply:
Hmm, apparently there's some timing issue with this.  I went ahead and just passed the ``$injector`` like I've been doing and had ``Pushes`` use that to get ``thNotify`` and ``$rootScope``.